# Исследование надежности заемщиков.

### Шаг 1. Обзор данных

In [1]:
import pandas as pd #импортируем библиотеку
data = pd.read_csv('/datasets/data.csv') #читаем файл и сохраняем в переменную
data.info()#получаем информации о файле
display(data.head(50))
display(data.tail(50))
#просматриваем первые и последние строчки таблицы
#просматриваем первые и последние строчки таблицы

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0   -1682.083438         47            начальное             3   
32         0   -4649.910832         34              СРЕДНЕЕ             1   
33         0   -1548.637544         48              среднее             1   
34         0   -4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0    -176.216688         33              среднее             1   
37         0   -6448.810860         43               ВЫСШЕЕ             0   
38         0    -597.881827         25               высшее             0   
39         0    -650.587796         31               высшее             0   
40         0   -1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0   -1257.496190         20  неоконченное высшее             2   
43         0   -4375.681384         43  неоконченное высшее             2   
44         1   -1362.041728         26              СРЕДНЕЕ             1   
45         1   -1039.451741         49              среднее             1   
46         2   -2262.712304         37              среднее             1   
47         1   -2689.137274         33               высшее             0   
48         0   -3341.067886         45              среднее             1   
49         0   -1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

children  days_employed  dob_years            education  education_id  \
21475         0  365213.306266         52              среднее             1   
21476         0   -2539.502794         52               высшее             0   
21477         0   -1628.432350         54              среднее             1   
21478         1    -921.332705         28               Высшее             0   
21479         0    -685.565994         26              среднее             1   
21480         0   -8799.354717         45               высшее             0   
21481         0    -829.519627         30  неоконченное высшее             2   
21482         1    -865.081118         25              среднее             1   
21483         2   -2917.035232         38            начальное             3   
21484         3    -968.718123         35              среднее             1   
21485         0  363996.449108         59               высшее             0   
21486         0   -1614.022809         36              СРЕДНЕЕ             1   
21487         0   -2376.789344         30              СРЕДНЕЕ             1   
21488         0    -354.755760         35               высшее             0   
21489         2            NaN         47              Среднее             1   
21490         2   -1867.006757         32              среднее             1   
21491        20    -173.954460         27              среднее             1   
21492         0   -1618.645283         54              среднее             1   
21493         1   -1214.858885         38              среднее             1   
21494         0    -960.095747         39              среднее             1   
21495         1            NaN         50              среднее             1   
21496         1    -759.680851         31              среднее             1   
21497         0            NaN         48               ВЫСШЕЕ             0   
21498         1   -1330.627998         32              среднее             1   
21499         0   -9929.015065         57              среднее             1   
21500         0    -578.082757         26  неоконченное высшее             2   
21501         0  334343.096304         57              среднее             1   
21502         1            NaN         42              среднее             1   
21503         0   -3096.881131         58              среднее             1   
21504         0  355235.728158         68              среднее             1   
21505         0  338904.866406         53              среднее             1   
21506         1   -1556.249906         33               ВЫСШЕЕ             0   
21507         1     -79.832064         32              среднее             1   
21508         0  386497.714078         62              среднее             1   
21509         0  362161.054124         59               высшее             0   
21510         2            NaN         28              среднее             1   
21511         0    -612.569129         29               высшее             0   
21512         0    -165.377752         26               высшее             0   
21513         0   -1166.216789         35              среднее             1   
21514         0    -280.469996         27  неоконченное высшее             2   
21515         1    -467.685130         28              среднее             1   
21516         0    -914.391429         42               высшее             0   
21517         0    -404.679034         42               высшее             0   
21518         0  373995.710838         59              СРЕДНЕЕ             1   
21519         1   -2351.431934         37       ученая степень             4   
21520         1   -4529.316663         43              среднее             1   
21521         0  343937.404131         67              среднее             1   
21522         1   -2113.346888         38              среднее             1   
21523         3   -3112.481705         38              среднее             1   
21524  

### Шаг 2.1 Заполнение пропусков

In [2]:
display(data.isna().sum())
data['total_income'] = data['total_income'].fillna(data['total_income'].median())# заполнили пропуски нужного столбца.
data.isna().sum() # проверили, устранили пропуски.
 

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

1)Пропущенные значения имеются сразу в двух столбцах и их одинаковое количество.

2)Пропущенные значения в каждом из столбцов с пропусками составляют  примерно 10% от всей таблицы.

3)Заполнить пропуски медианным значением — лучшее решение для количественных переменных, потому что: 1) нам необохдимо заполнить пропуски для анализа 2) Cреднее значение не явялется лучшим решением, так как медиана дает более реалистичную оценку центрального значения.

4)Причина отсутвия 10% данных столбца "income_type" связана с недостующими данными в столбце "days_employed", также мы можем увидеть, что в "days_employed" некотрве данные имеют невомозжные значения. 

### Шаг 2.2 Проверка данных на аномалии и исправления.

In [3]:
data['days_employed'] = data['days_employed'].abs() #изменили отрицательные значения функией abs()
data['days_employed'] = data['days_employed'].fillna(data['days_employed'].median())# заполняем пропуски медианными значениями
#display(data['dob_years'].value_counts()) # возраст  некоторых клиентов равен 0, что ошибка, заменем медианным значением.
data.loc[data['dob_years'] == 0,'dob_years'] = data['dob_years'].median()
data['dob_years'] = data['dob_years'].astype('int')
#display(data['dob_years'].value_counts())# проверили,изменилось.
data['gender'].value_counts()# выявилась одна аномалия, однако она нам не помешает при анализе, не будем ее изменять.
data['children'].value_counts()# появилось два анамальных значения, вероятнее это ошибка заполнявшего, попробуем устранить
data.loc[data['children'] == 20,'children'] = 2
data.loc[data['children'] == -1,'children'] = 1
display(data.isna().sum())
data.head(50) # проверяем изменения


children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0    2194.220567         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              СРЕДНЕЕ             1   
20         1    1311.604166         36              среднее             1   
21         1     253.685166         33              среднее             1   
22         1    1766.644138         24              среднее             1   
23         0     272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0    2194.220567         41              среднее             1   
27         0     529.191635         28               высшее             0   
28         1     717.274324         26               высшее             0   
29         0    2194.220567         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0    1682.083438         47            начальное             3   
32         0    4649.910832         34              СРЕДНЕЕ             1   
33         0    1548.637544         48              среднее             1   
34         0    4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0     176.216688         33              среднее             1   
37         0    6448.810860         43               ВЫСШЕЕ             0   
38         0     597.881827         25               высшее             0   
39         0     650.587796         31               высшее             0   
40         0    1030.362606         30              среднее             1   
41         0    2194.220567         50              среднее             1   
42         0    1257.496190         20  неоконченное высшее             2   
43         0    4375.681384         43  неоконченное высшее             2   
44         1    1362.041728         26              СРЕДНЕЕ             1   
45         1    1039.451741         49              среднее             1   
46         2    2262.712304         37              среднее             1   
47         1    2689.137274         33               высшее             0   
48         0    3341.067886         45              среднее             1   
49         0    1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

Причинами  появления таких данных могут появится из-за человеческих факторы, либо в результате технических ошибок(например перенос данных). 
Также, помимо отрицательных значений и пропусков, в некоторых строчках появляются значения, которые невозмможны. Однако эти данные никак не повлияют на наш анализ, поэтому мы не будем трогать их.


### Шаг 2.3. Изменение типов данных.

In [4]:
data['total_income'] = data['total_income'].astype('int') 
display(data)
data.info()
# заменили тип данных на целочисленный и проверили изменения


children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   Среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21520        224791             операции с жильем  
21521        155999          сделка с автомобилем  
21522         89672                  недвижимость  
21523        244093  на покупку своего автомобиля  
21524         82047         на покупку автомобиля  

[21525 rows x 12 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


### Шаг 2.4. Удаление дубликатов.

In [5]:
display(data.duplicated().sum())
display(data['education'].value_counts())# необходимо обработать
#display(data['income_type'].value_counts())# в этом столбце все в порядке
#display(data['purpose'].value_counts())# много значений, их мы обработаем чуть позже
#data['family_status'].value_counts()
data['family_status'] = data['family_status'].str.lower()
data['education'] = data['education'].str.lower()# привел к нижнему регистру
display(data['education'].value_counts())#исправили
data = data.drop_duplicates().reset_index(drop=True)
display(data.duplicated().sum())
data.head(50)



55

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

0

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              среднее             1   
8          2    6929.865299         35               высшее             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0    2194.220567         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              среднее             1   
20         1    1311.604166         36              среднее             1   
21         1     253.685166         33              среднее             1   
22         1    1766.644138         24              среднее             1   
23         0     272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0    2194.220567         41              среднее             1   
27         0     529.191635         28               высшее             0   
28         1     717.274324         26               высшее             0   
29         0    2194.220567         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0    1682.083438         47            начальное             3   
32         0    4649.910832         34              среднее             1   
33         0    1548.637544         48              среднее             1   
34         0    4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0     176.216688         33              среднее             1   
37         0    6448.810860         43               высшее             0   
38         0     597.881827         25               высшее             0   
39         0     650.587796         31               высшее             0   
40         0    1030.362606         30              среднее             1   
41         0    2194.220567         50              среднее             1   
42         0    1257.496190         20  неоконченное высшее             2   
43         0    4375.681384         43  неоконченное высшее             2   
44         1    1362.041728         26              среднее             1   
45         1    1039.451741         49              среднее             1   
46         2    2262.712304         37              среднее             1   
47         1    2689.137274         33               высшее             0   
48         0    3341.067886         45              среднее             1   
49         0    1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

1) Для удаления дубликатов мы воспользовались: метод drop_duplicates для явных дубликатов, затем value_counts для уникальных значений и в после этого пришлось ручнымм метод переводить строки к нижнему регистру. Таким образом, мы избавились от всех дубликатов и улучшили таблицу
2) Главной причиной появления дубликатов в данной ситуации - человеческий фактор, некоторые значения были написаны по-разному, хоть и имели один и тот же смысл.

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [6]:
education_info = data[['education','education_id']]
family_status_info = data[['family_status','family_status_id']]
# создали два новых датафрейма
display(education_info.value_counts())
display(family_status_info.value_counts())
try:
    data = data.drop(columns = ['education','family_status'])
except:
    display('Ошибка, примените другой способ')
# удалили из исходного датафрейма столбцы education и family_status
print(education_info.duplicated().sum())
print(family_status_info.duplicated().sum())
education_info = education_info.drop_duplicates().reset_index(drop=True)
family_status_info = family_status_info.drop_duplicates().reset_index(drop=True)
print(education_info.duplicated().sum())
print(family_status_info.duplicated().sum())
display(data.head(30))
display(data.tail(30))



education            education_id
среднее              1               15171
высшее               0                5250
неоконченное высшее  2                 744
начальное            3                 282
ученая степень       4                   6
dtype: int64

family_status          family_status_id
женат / замужем        0                   12339
гражданский брак       1                    4150
не женат / не замужем  4                    2810
в разводе              3                    1195
вдовец / вдова         2                     959
dtype: int64

21448
21448
0
0


children  days_employed  dob_years  education_id  family_status_id gender  \
0          1    8437.673028         42             0                 0      F   
1          1    4024.803754         36             1                 0      F   
2          0    5623.422610         33             1                 0      M   
3          3    4124.747207         32             1                 0      M   
4          0  340266.072047         53             1                 1      F   
5          0     926.185831         27             0                 1      M   
6          0    2879.202052         43             0                 0      F   
7          0     152.779569         50             1                 0      M   
8          2    6929.865299         35             0                 1      F   
9          0    2188.756445         41             1                 0      M   
10         2    4171.483647         36             0                 0      M   
11         0     792.701887         40             1                 0      F   
12         0    2194.220567         65             1                 1      M   
13         0    1846.641941         54             2                 0      F   
14         0    1844.956182         56             0                 1      F   
15         1     972.364419         26             1                 0      F   
16         0    1719.934226         35             1                 0      F   
17         0    2369.999720         33             0                 1      M   
18         0  400281.136913         53             1                 2      F   
19         0   10038.818549         48             1                 3      F   
20         1    1311.604166         36             1                 0      M   
21         1     253.685166         33             1                 1      F   
22         1    1766.644138         24             1                 1      F   
23         0     272.981385         21             0                 1      M   
24         1  338551.952911         57             1                 4      F   
25         0  363548.489348         67             1                 0      M   
26         0    2194.220567         41             1                 0      M   
27         0     529.191635         28             0                 0      M   
28         1     717.274324         26             0                 0      F   
29         0    2194.220567         63             1                 4      F   

    income_type  debt  total_income                                 purpose  
0     сотрудник     0        253875                           покупка жилья  
1     сотрудник     0        112080                 приобретение автомобиля  
2     сотрудник     0        145885                           покупка жилья  
3     сотрудник     0        267628              дополнительное образование  
4     пенсионер     0        158616                         сыграть свадьбу  
5     компаньон     0        255763                           покупка жилья  
6     компаньон     0        240525                       операции с жильем  
7     сотрудник     0        135823                             образование  
8     сотрудник     0         95856                   на проведение свадьбы  
9     сотрудник     0        144425                 покупка жилья для семьи  
10    компаньон     0        113943                    покупка недвижимости  
11    сотрудник     0         77069       покупка коммерческой недвижимости  
12    пенсионер     0        145017                         сыграть свадьбу  
13    сотрудник     0        130458                 приобретение автомобиля  
14    компаньон     1        165127              покупка жилой недвижимости  
15    сотрудник     0        116820  строительство собственной недвижимости  
16    сотрудник     0        289202                            недвижимость  
17    сотрудник     0         90410              строительство недвижимости  
18    пенс

children  days_employed  dob_years  education_id  family_status_id  \
21423         1    2194.220567         50             1                 1   
21424         1     759.680851         31             1                 0   
21425         0    2194.220567         48             0                 0   
21426         1    1330.627998         32             1                 1   
21427         0    9929.015065         57             1                 1   
21428         0     578.082757         26             2                 4   
21429         0  334343.096304         57             1                 0   
21430         1    2194.220567         42             1                 0   
21431         0    3096.881131         58             1                 0   
21432         0  355235.728158         68             1                 0   
21433         0  338904.866406         53             1                 1   
21434         1    1556.249906         33             0                 1   
21435         1      79.832064         32             1                 1   
21436         0  386497.714078         62             1                 0   
21437         0  362161.054124         59             0                 0   
21438         2    2194.220567         28             1                 0   
21439         0     612.569129         29             0                 1   
21440         0     165.377752         26             0                 4   
21441         0    1166.216789         35             1                 0   
21442         0     280.469996         27             2                 4   
21443         1     467.685130         28             1                 0   
21444         0     914.391429         42             0                 0   
21445         0     404.679034         42             0                 1   
21446         0  373995.710838         59             1                 0   
21447         1    2351.431934         37             4                 3   
21448         1    4529.316663         43             1                 1   
21449         0  343937.404131         67             1                 0   
21450         1    2113.346888         38             1                 1   
21451         3    3112.481705         38             1                 0   
21452         2    1984.507589         40             1                 0   

      gender  income_type  debt  total_income  \
21423      F    сотрудник     0        145017   
21424      F    сотрудник     0        119392   
21425      F    компаньон     0        145017   
21426      M    сотрудник     0        240767   
21427      M    компаньон     0        157553   
21428      M    компаньон     0         77813   
21429      F    пенсионер     0         86232   
21430      F    сотрудник     0        145017   
21431      F    сотрудник     0        264250   
21432      F    пенсионер     0         80566   
21433      M    пенсионер     0         75439   
21434      F    сотрудник     0        145541   
21435      F  госслужащий     0         98180   
21436      M    пенсионер     0         72638   
21437      M    пенсионер     0         73029   
21438      F    сотрудник     0        145017   
21439      F    сотрудник     1        140068   
21440      M    компаньон     0        147301   
21441      F    сотрудник     0        250986   
21442      M    компаньон     0        355988   
21443      F    сотрудник     1        109486   
21444      F    компаньон     0        322807   
21445      F    компаньон     0        178059   
21446      F    пенсионер     0        153864   
21447      M    сотрудник     0        115949   
21448      F    компаньон     0        224791   
21449      F    пенсионер     0        155999   
21450      M    сотрудник     1         89672   
21451      M    сотрудник     1        244093   
21452      F    сотрудник     0         82047   

                                     purpose  
21423                                свадьба  
21424

In [7]:
# Код ревьюера

try:
    1 / 0
except:
    print('Произошла какая-то ошибка')
    
try:
    1 / 0
except ZeroDivisionError:
    print('Произошло деление на 0')
    
# 1 / 0

Произошла какая-то ошибка
Произошло деление на 0


### Шаг 2.6. Категоризация дохода.

In [8]:
def total_income_category(row):
    total_income = row['total_income']
    if total_income <= 30000:
        return 'E'
    if total_income <= 50000:
        return 'D'
    if total_income <= 200000:
        return 'C'
    if total_income <= 1000000:
        return 'B'
    return 'A'
data['total_income_category'] = data.apply(total_income_category, axis=1)
display(data['total_income_category'].value_counts())
data
# создали функцию для распределения на категории в зависимости от уровня дохода и добавили новый столбец

C    16015
B     5041
D      350
A       25
E       22
Name: total_income_category, dtype: int64

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  340266.072047         53             1                 1   
...         ...            ...        ...           ...               ...   
21448         1    4529.316663         43             1                 1   
21449         0  343937.404131         67             1                 0   
21450         1    2113.346888         38             1                 1   
21451         3    3112.481705         38             1                 0   
21452         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21448      F   компаньон     0        224791             операции с жильем   
21449      F   пенсионер     0        155999          сделка с автомобилем   
21450      M   сотрудник     1         89672                  недвижимость   
21451      M   сотрудник     1        244093  на покупку своего автомобиля   
21452      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category  
0                         B  
1                         C  
2                         C  
3                         B  
4                         C  
...                     ...  
21448                     B  
21449                     C  
21450                     C  
21451                     B  
21452                     C  

[21453 rows x 11 columns]

### Шаг 2.7. Категоризация целей кредита.

In [9]:
data['purpose'].value_counts().sum()#проверяем количество уникальных значений в столбце purpose
def purpose_category(purpose):
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    if 'недвижимост'in purpose or 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образов' in purpose:
        return 'получение образования'
data['purpose_category'] = data['purpose'].apply(purpose_category)
#создали функцию, которая категоризирует цели кредита.
display(data['purpose_category'].value_counts())
display(data['purpose_category'].value_counts().sum())# количество уникальных значений в столбце purpose_category сходится со столбцом purpose
data

    
    

операции с недвижимостью    10811
операции с автомобилем       4306
получение образования        4013
проведение свадьбы           2323
Name: purpose_category, dtype: int64

21453

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  340266.072047         53             1                 1   
...         ...            ...        ...           ...               ...   
21448         1    4529.316663         43             1                 1   
21449         0  343937.404131         67             1                 0   
21450         1    2113.346888         38             1                 1   
21451         3    3112.481705         38             1                 0   
21452         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21448      F   компаньон     0        224791             операции с жильем   
21449      F   пенсионер     0        155999          сделка с автомобилем   
21450      M   сотрудник     1         89672                  недвижимость   
21451      M   сотрудник     1        244093  на покупку своего автомобиля   
21452      F   сотрудник     0         82047         на покупку автомобиля   

      total_income_category          purpose_category  
0                         B  операции с недвижимостью  
1                         C    операции с автомобилем  
2                         C  операции с недвижимостью  
3                         B     получение образования  
4                         C        проведение свадьбы  
...                     ...                       ...  
21448                     B  операции с недвижимостью  
21449                     C    операции с автомобилем  
21450                     C  операции с недвижимостью  
21451                     B    операции с автомобилем  
21452                     C    операции с автомобилем  

[21453 rows x 12 columns]

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h5>

Верно разбиты клиенты по категориям целей получения кредита!

</div>

### Ответы на вопросы.

Выводим среднее значение функцией mean(). dept = 0 - не имел задолженности, debt = 1 - имел задолдонить. Таким образом, категории в которых среднее значение ближе к 1 более склонны к невыплате кредита в срок.

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [10]:
data_children_pivot = data.pivot_table(index='children', values = 'debt', aggfunc = 'mean')
data_children_pivot

debt
children          
0         0.075444
1         0.091658
2         0.094925
3         0.081818
4         0.097561
5         0.000000

##### Вывод 1:  Исходя из результата, можно сделать вывод, что люди с 5 детьми почти всегда выплачивают кредит в срок.  Люди с 1,2 и 4 детьми наиболее расположены к неуплате в срок.

##### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [11]:
data_family_pivot = data.pivot_table(index='family_status_id', values = 'debt', aggfunc = 'mean')
data_family_pivot
#построим сводную таблицу, чем ближе значения столюбца debt к 1, тем большее количество людей не выплачивают кредит вовремя

debt
family_status_id          
0                 0.075452
1                 0.093494
2                 0.065693
3                 0.071130
4                 0.097509

##### Вывод 2:  Таким образом, исходя из данных можно сделать вывод, что люди неженатые /незамужемужние и в гражданском браке чаще остальных не выплачивают кредит в срок. Вдовцы и вдовы наооборот имеет самый низкий показатель и выплачивают  кредит вовремя чаще всех.

##### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [12]:
data_income_pivot = data.pivot_table(index = 'total_income_category', values = 'debt', aggfunc = 'mean')
data_income_pivot
#построим сводную таблицу, чем ближе значения столюбца debt к 1, тем большее количество людей из определенной категории не выплачивают кредит вовремя

debt
total_income_category          
A                      0.080000
B                      0.070621
C                      0.084920
D                      0.060000
E                      0.090909

##### Вывод 3:  Таким образом,  наибольшее количетсво людей, которые не выплачивают кредит вовремя относятся к категории людей чей доход  до 30 тыс..  Самые ответственные  люди относятся к категории от 30 то 50 тыс.. 

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [13]:
data_purpose_pivot = data.pivot_table(index = 'purpose_category', values = 'debt', aggfunc = 'mean')
data_purpose_pivot
#построим сводную таблицу, чем ближе значения столюбца debt к 1, тем большее количество людей не выплачивают на определеннную категорию  кредит  вовремя

debt
purpose_category                  
операции с автомобилем    0.093590
операции с недвижимостью  0.072334
получение образования     0.092200
проведение свадьбы        0.080069

##### Вывод 4:  По таблице можно сделать вывод, что  на операции с автомобилем и получение образования, полученные кредиты возвращаются позже срока гораздо больше двух других целей.

## Общий вывод:
Данные были предобработаны следующим образом:заполнили пропуски, удалили дубликаты, исправили аномальные значения, создали недостающие столбцы. С помощью этих действий смогли выполнить поставленную задачу и проанализировать данные и сделать выводы о том, влияет ли семенное положение, количество детей, заработок и цель займа на погашение кредита в срок.
Таким образом, наилучшие категории для выдачи кредитов являются:
1. Семьи с 5 детьми, c вероятностью 0.000000 не выплачивают кредит вовремя.
2. Вдовцы и Вдовы, c вероятностью 0.065693 не выплачивают кредит вовремя.
3. Люди с достатком в размере 30-50 тыс., c вероятностью 0.060000 не выплачивают кредит вовремя.
4. Кредит на приобретение недвижимости и проведения свадеб, c вероятностью 0.072334 и 0.080069 не выплачивают кредит вовремя.

Наихудшие категории для выдачи кредитов являются:
1. Люди с 1,2 и 4 детьми, c вероятностью 0.091658, 0.094925 и 0.097561 не выплачивают кредит вовремя.
2. Неженатые /незамужемужние и в гражданском браке, c вероятностью 0.093494 и 0.097509 не выплачивают кредит вовремя.
3. Люди с достаком до 30 тыс., c вероятностью 0.090909 не выплачивают кредит вовремя.
4. Кредит на операции с автомобилем	и получения образования, c вероятностью 0.093590 и 0.092200 не выплачивают кредит вовремя.

